Documentation/contributing/development/statedb.rst

```diff
@@ -0,0 +1,443 @@
+^^^^^^^^
+
+BPF maps can be reconciled with the operations returned by ``bpf.NewMapOps``.
+The target object needs to implement the ``BinaryKey`` and ``BinaryValue`` to
+construct the BPF key and value respectively. These can either construct the binary
+value on the fly, or reference a struct defining the value. The example below uses
+a struct as this is the prevalent style in Cilium.
+
+.. code-block:: go
+
+  // MyKey defines the raw BPF key
+  type MyKey struct { ... }
+  // MyValue defines the raw BPF key
+  type MyValue struct { ... }
+
+  type MyObject struct {
+    Key MyKey
+	Value MyValue
+    Status reconciler.Status
+  }
+
+  func (m *MyObject) BinaryKey() encoding.BinaryMarshaler {
+    return bpf.StructBinaryMarshaler{&m.Key}
+  }
+  func (m *MyObject) BinaryValue() encoding.BinaryMarshaler {
+    return bpf.StructBinaryMarshaler{&m.Value}
+  }
+
+  func registerReconciler(params reconciler.Params, objs statedb.RWTable[*MyObject], m *bpf.Map) error {  
+    ops := bpf.NewMapOps[*MyObject](m)
+    _, err := reconciler.Register(
+      params,
+      objs,
+      func(obj *MyObject) *MyObject { return obj },
+      func(obj *MyObject, s reconciler.Status) *MyObject {
+        obj.Status = obj
+        return obj
+      },
+      func(obj *MyObject) reconciler.Status {
+        return e.Status
+      },
+      ops,
+      nil,
+    )
+    return err
+  }
+
+For a real-world example see ``pkg/maps/bwmap/cell.go``.
+
+Metrics
+~~~~~~~
+
+Metrics are available for both StateDB and the reconciler, but they are disabled
+by default due to their fine granularity. These are defined in ``pkg/hive/statedb_metrics.go``
+and ``pkg/hive/reconciler_metrics.go``. As this documentation is manually maintained, it may
+be out-of-date, so if things are not working, check the source code.
+
+The metrics can be enabled by adding them to the helm ``prometheus.metrics`` option with
+the syntax ``+cilium_<name>``, where ``<name>`` is the name of the metric in the table below.
+For example, here is how to turn on all the metrics:
+
+.. code-block:: yaml
+
+  prometheus:
+    enabled: true
+    metrics:
+    - +cilium_statedb_write_txn_duration_seconds
+    - +cilium_statedb_write_txn_acquisition_seconds
+    - +cilium_statedb_table_contention_seconds
+    - +cilium_statedb_table_objects
+    - +cilium_statedb_table_revision
+    - +cilium_statedb_table_delete_trackers
+    - +cilium_statedb_table_graveyard_objects
+    - +cilium_statedb_table_graveyard_low_watermark
+    - +cilium_statedb_table_graveyard_cleaning_duration_seconds
+    - +cilium_reconciler_count
+    - +cilium_reconciler_duration_seconds
+    - +cilium_reconciler_errors_total
+    - +cilium_reconciler_errors_current
+    - +cilium_reconciler_prune_count
+    - +cilium_reconciler_prune_errors_total
+    - +cilium_reconciler_prune_duration_seconds
+
+These are still under development and the metric names may change. We also do not
+as of yet have a grafana dashboard template for these.
```

Comment:
“as of yet” is tricky in docs. How about creating an issue to track this and pointing to it in the docs?

Reply:
I'll drop this. It's not yet entirely clear if/how/when we'll do a dashboard for the reconciler metrics.